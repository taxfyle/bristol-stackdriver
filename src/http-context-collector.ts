import { IncomingMessage, ServerResponse } from 'http'
import { StackdriverHttpContext } from '.'

/**
 * Collects HTTP context info from Node's HTTP Request and Response objects.
 * @param req
 * @param res
 */
export function collectHttpContext(
  req: IncomingMessage,
  res: ServerResponse
): StackdriverHttpContext {
  const headers = req.headers || /* istanbul ignore next */ {}
  return {
    method: req.method!,
    url: req.url!,
    userAgent: headers['user-agent'] as string,
    referrer: headers['referer'] as string,
    responseStatusCode: res.statusCode,
    remoteIp: getRemoteIp(req)
  }
}

/**
 * Gets the remote IP, taking into account `X-Forwarded-For` headers.
 * @param req
 */
function getRemoteIp(req: IncomingMessage): string {
  /* istanbul ignore else */
  if (req.headers) {
    const header = req.headers['x-forwarded-for']
    if (typeof header !== 'undefined') {
      return header as string
    }
  }

  /* istanbul ignore else */
  /* tslint:disable-next-line:strict-type-predicates */
  if (req.connection && typeof req.connection === 'object') {
    return req.connection.remoteAddress /* istanbul ignore next */ || ''
  }

  /* istanbul ignore next */
  return ''
}
