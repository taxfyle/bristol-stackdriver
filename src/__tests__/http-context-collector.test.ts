import * as http from 'http'
import axios from 'axios'
import { collectHttpContext } from '../http-context-collector'
import { StackdriverHttpContext } from '..'

jest.setTimeout(20000)

describe('http context collector', () => {
  it('collects the correct context', async () => {
    const contexts: Array<StackdriverHttpContext> = []
    const server = http
      .createServer((req, res) => {
        res.statusCode = 200
        contexts.push(collectHttpContext(req, res))
        res.write('Waddup')
        res.end()
      })
      .listen()

    try {
      const addr = server.address()
      const url = `http://127.0.0.1:${addr.port}`
      const client = axios.create({ baseURL: url })
      await client.get('/IAmTheDerg/TheBigBadDerg')
      await client.post(
        '/WhatDoWeSayToDeath',
        { answer: 'Not Today' },
        {
          headers: {
            referer: 'ya mum',
            'user-agent': 'austin powers',
            'x-forwarded-for': '1.2.3.4'
          }
        }
      )

      expect(contexts).toMatchSnapshot()
    } finally {
      server.close()
    }
  })
})
