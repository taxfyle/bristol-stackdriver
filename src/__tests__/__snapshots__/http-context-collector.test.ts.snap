// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`http context collector collects the correct context 1`] = `
Array [
  Object {
    "method": "GET",
    "referrer": undefined,
    "remoteIp": "::ffff:127.0.0.1",
    "responseStatusCode": 200,
    "url": "/IAmTheDerg/TheBigBadDerg",
    "userAgent": "axios/0.19.0",
  },
  Object {
    "method": "POST",
    "referrer": "ya mum",
    "remoteIp": "1.2.3.4",
    "responseStatusCode": 200,
    "url": "/WhatDoWeSayToDeath",
    "userAgent": "austin powers",
  },
]
`;
